reactivesocket-client/src/main/java/io/reactivesocket/client/LoadBalancer.java

```diff
@@ -121,18 +179,59 @@ public LoadBalancer(Publisher<List<ReactiveSocketFactory<SocketAddress>>> factor
     }
 
     private synchronized void addSockets(int numberOfNewSocket) {
-        activeFactories.entrySet()
-            .stream()
-            // available factories that don't map to an already established socket
-            .filter(e -> !activeSockets.containsKey(e.getKey()))
-            .map(e -> e.getValue())
-            .filter(factory -> factory.availability() > 0.0)
-            .sorted((a, b) -> -Double.compare(a.availability(), b.availability()))
-            .limit(numberOfNewSocket)
-            .forEach(factory -> {
-                pendingSockets += 1;
-                factory.apply().subscribe(new SocketAdder(factory.remote()));
-            });
+        int n = numberOfNewSocket;
+        if (n > activeFactories.size()) {
+            n = activeFactories.size();
+            logger.info("addSockets(" + numberOfNewSocket
+                + ") restricted by the number of factories, i.e. addSockets(" + n + ")");
+        }
+
+        Random rng = ThreadLocalRandom.current();
+        while (n > 0) {
+            int size = activeFactories.size();
+            if (size == 1) {
+                ReactiveSocketFactory<T> factory = activeFactories.get(0);
+                if (factory.availability() > 0.0) {
+                    activeFactories.remove(0);
+                    pendingSockets++;
+                    factory.apply().subscribe(new SocketAdder(factory));
+                }
+                break;
+            }
+            ReactiveSocketFactory<T> factory0 = null;
+            ReactiveSocketFactory<T> factory1 = null;
+            int i0 = 0;
+            int i1 = 0;
+            for (int i = 0; i < EFFORT; i++) {
+                i0 = rng.nextInt(size);
+                i1 = rng.nextInt(size - 1);
+                if (i1 >= i0) {
```

Comment:
I want to find two different random numbers here in [0, n].
I pick the first one randomly, let's call it `i0`.
`[.....i0.....]`
I want to pick another one `i1` which has to be different from x, so I pick a random number in [0, n-1]
`[.....i0.....]`
`[...........]`

Two cases here:
either `i1` is < `i0` and that's fine
`[.....i0.....]`
`[.i1..i0.....]`

or `i1` is >= `i0` then I need to increment `i1` because of the spot that `i0` is already taking.
`[.....i0.....]`
`[.....i0..i1.]`

If you don't do that, and only check for the case where you pick the same number twice, you bias the second random number to `i0 + 1` with a 2/(n-1) probability. (There's two cases where `i1 = i0+1`, when you secondly pick `i0` and `i0+1`)